src/language/dfdl.ts

```diff
@@ -0,0 +1,1328 @@
+  // 			if(wholeLine.endsWith("'")) {
+  // 				vscode.window.activeTextEditor?.insertSnippet(new vscode.SnippetString("$1'$0"), position);
+  // 			}
+  // 			console.log('endSingleQuoteProvider returning undefined');
+  // 			return undefined;
+  // 		}
+  // 	},
+  // 	"'"
+
+  const endSingleBraceProvider =
+    vscode.languages.registerCompletionItemProvider(
+      'dfdl',
+      {
+        provideCompletionItems(
+          document: vscode.TextDocument,
+          position: vscode.Position
+        ) {
+          //				const posMinusOne = document.lineAt(position).text.substr(0, position.character - 1);
+          const wholeLine = document
+            .lineAt(position)
+            .text.substr(0, position.character)
+          if (wholeLine.includes('dfdl:length="{')) {
+            vscode.window.activeTextEditor?.insertSnippet(
+              new vscode.SnippetString('$1}$0'),
+              position
+            )
+          }
+          return undefined
+        },
+      },
+      '{'
+    )
+
+  function getDefinedTypes(document: vscode.TextDocument) {
+    var additionalTypes = ''
+    var lineNum = 0
+    const lineCount = document.lineCount
+    while (lineNum !== lineCount) {
+      const wholeLine = document
+        .lineAt(lineNum)
+        .text.substring(0, document.lineAt(lineNum).range.end.character)
+      // console.log('getDefinedTypes - wholeLine: ' + wholeLine);
+      if (
+        wholeLine.includes('xs:simpleType Name=') ||
+        wholeLine.includes('xs:complexType Name=')
+      ) {
+        var startPos = wholeLine.indexOf('"', 0)
+        var endPos = wholeLine.indexOf('"', startPos + 1)
+        var newType = wholeLine.substring(startPos + 1, endPos)
+        additionalTypes = String(additionalTypes + ',' + newType)
+      }
+      ++lineNum
+    }
+    return additionalTypes
+  }
+
+  function getDefinedVariables(document: vscode.TextDocument) {
+    var additionalTypes = ''
+    var lineNum = 0
+    var itemCnt = 0
+    const lineCount = document.lineCount
+    while (lineNum !== lineCount) {
+      const wholeLine = document
+        .lineAt(lineNum)
+        .text.substring(0, document.lineAt(lineNum).range.end.character)
+      // console.log('getDefinedTypes - wholeLine: ' + wholeLine);
+      if (wholeLine.includes('dfdl:defineVariable name=')) {
+        var startPos = wholeLine.indexOf('"', 0)
+        var endPos = wholeLine.indexOf('"', startPos + 1)
+        var newType = wholeLine.substring(startPos + 1, endPos)
+        if (itemCnt === 0) {
+          additionalTypes = newType
+          ++itemCnt
+        } else {
+          additionalTypes = String(additionalTypes + ',' + newType)
+          ++itemCnt
+        }
+      }
+      ++lineNum
+    }
+    return additionalTypes
+  }
+
+  context.subscriptions.push(
+    elementCompletionProvider,
+    attributeCompletionProvider,
+    closeElementGtProvider,
+    closeElementSlashProvider,
+    endSingleBraceProvider
+  )
+}
```

Comment:
@rthomas320

Reply:
@stevedlawrence 
So we can have our extension requires other so we could require this for the xml intellisense but for being able to specific DFDL things we will need to do what we do now and make the custom items as that is what the extension you linked actually does as well just in a bit of a different way.

So the other way of doing it is creating a languageServer. This is also pretty involved to do and doesn't look as straight forward when you look at the code. I believe this would at least be a good start, then we can make an issue to create/look into creating a languageServer. Does that sounds good?

Reply:
Initially we tried using the Red XML extension for auto completion. The extension uses an XML catalog file similar to the XML catalog file that is described on the wiki page for using eclipse with daffodil.  The extension worked to an extent, but Boyd wanted the auto completion to be context sensitive and return only specific items based on the current context.  The Redhat extension couldn't do that just by supplying an XML catalog.

Reply:
By context, do you mean things like if `dfdl:lengthKind="pattern"` is used, then `dfdl:length` won't be provided as an auto completion option because only `dfdl:lengthPattern` is valid? Makes sense, that kind of information isn't described in the XSD for DFDL.

I'm fine with the snippet approach if it seems to be working. It feels like it lacks context in some cases, but maybe that can be overcome. Happy to see how it works out. And if we do run into limitations, perhaps we could investigate the feasibility of the language server and if it makes things any better without too much complication.